pwnlib/gdb_api_bridge.py

```diff
@@ -89,6 +89,18 @@ def stop(self):
             return Breakpoint(*args, **kwargs)
         return gdb.Breakpoint(*args, **kwargs)
 
+    def exposed_set_finish_breakpoint(self, client, has_stop, has_out_of_scope, *args, **kwargs):
+        """Create a finish breakpoint and connect it with the client-side mirror."""
+        class FinishBreakpoint(gdb.FinishBreakpoint):
+            def stop(self):
+                if has_stop:
+                    return client.stop()
+                return True
+            def out_of_scope(self):
+                if has_out_of_scope:
```

Comment:
For what is worth, those checks do not have to be evaluated inside:
<img width="497" alt="image" src="https://user-images.githubusercontent.com/10009354/196027067-5861fe05-6800-477e-b876-e0052a400c1c.png">

Reply:
(I am not saying that is the way we want to go, but if performance matters here anyhow, e.g. those callbacks may be called many many times then maybe we do want, idk)

Reply:
I agree with you that the checks can be evaluated outside (I didn't know Python supports such a syntax).

However, both the `stop` and `out_of_scope` callbacks of the finish breakpoints will only be called at most once. So I don't think the performance matters here.

Nevertheless, I prefer to evaluate the checks outside, as you have mentioned. If this is the way to go, I will add a commit to fix this.

Reply:
Isn't this "once per hit"? I mean, it may not matter if you do set or hit the bp just once, but if you do it in a scripted way, I guess it may matter.

Reply:
Yes, it is evaluated once per hit. Since the finish breakpoint will be deleted by GDB after hit, the evaluations will only be done once. We can either evaluate it when the breakpoint is hit (i.e., evaluate inside), or evaluate it while setting the breakpoint (i.e., evaluate outside).

Reply:
Just added a commit to evaluate the variables outside.